schema/ref.go

```diff
@@ -80,22 +95,43 @@ func (r *RefSchema) Display() Display {
 	return r.DisplayValue
 }
 
-func (r *RefSchema) ApplyScope(scope Scope) {
-
+func (r *RefSchema) ApplyScope(scope Scope, namespace string) {
+	if namespace != r.ObjectNamespace {
+		return // The scope does not apply to this reference.
+	}
 	objects := scope.Objects()
 	referencedObject, ok := objects[r.IDValue]
 	if !ok {
 		panic(BadArgumentError{
-			Message: fmt.Sprintf("Referenced object '%s' not found in scope", r.IDValue),
+			Message: fmt.Sprintf("Referenced object '%s' not found in scope with namespace %q", r.IDValue, namespace),
 		})
 	}
 	r.referencedObjectCache = referencedObject
 }
 
+func (r *RefSchema) ValidateReferences() error {
+	if r.referencedObjectCache != nil {
+		return nil // Success
+	}
+	// The only way, unless there is a bug, for it to get here is if ApplyScope was not called with the
+	// correct namespace, or if the code disregards the error returned by ApplyScope. ApplyScope should
```

Comment:
This sounds like "the only way to get here is if there's a bug, unless there's a bug". Or are "was not called with the correct namespace" and "disregards the error returned" not bugs??? 😁 

Also, if there's what would appear to be an internal Arca logic error, do we really want to return a `BadArgumentError` vs a panic?

Reply:
This is the only place that's designed to validate it. Users can hit this with a bad workflow schema, which will happen a lot. This is designed to be called before all of the other functions that reference the object cache. If this isn't called (or isn't properly handled), it will get to the other calls, which will panic instead of just return an error.
You could say that there are two types of bugs. Engine and SDK bugs, and workflow bugs. Users will hit workflows "bugs" all the time due to mistakes in the schemas written. This function is designed to point out workflow bugs under the assumption that there are no bugs in the SDK or engine that create misleading errors.

Reply:
If you expect this to be seen by end users, can we come up with a more direct message tells them what the workflow file did wrong?

Reply:
The error message given is "Ref object reference missing its link to object with ID %q in namespace %q. Is that a valid namespace?"
So that's about as much context as we can give them internally. Externally, I add even more context: https://github.com/arcalot/arcaflow-engine/pull/167/files#diff-786b3ca53611bc6c73c4444dc75509072a95df7ac2015ebe0db97ca37c30f20aR356

Reply:
Since this isn't the primary error presented to a user, I withdraw the comments; although I'm not entirely convinced that a user encountering this will end up with sufficient information to figure out what they did, that's a separate issue.

I'm still not crazy about the code comment. It sounds like either there's an internal engine error or maybe the user referenced a symbol with the wrong namespace: neither of which are clearly represented in the comment. But I guess I'll leave this one be.

Reply:
Jared and I went through the exercise of generating this error last night (thanks @jaredoconnell!).

As the code currently stands, if the namespace ID (name) is bad, the message will print it out along with the valid names; if the reference is bad, we die elsewhere (and, sadly, with a traceback...) but with a message which points out that the reference (by name) is not available in the named scope/namespace.  So, I think there is enough information available, either way, for the user to track the problem to the appropriate place in the workflow and to be able fix it.